raiden/network/proxies/discovery.py

```diff
@@ -42,12 +44,15 @@ def __init__(
 
         check_address_has_code(jsonrpc_client, discovery_address, 'Discovery')
 
-        is_valid_version = compare_versions(
-            proxy.contract.functions.contract_version().call(),
-            EXPECTED_CONTRACTS_VERSION,
-        )
-        if not is_valid_version:
-            raise ContractVersionMismatch('Incompatible ABI for Discovery')
+        try:
+            is_valid_version = compare_versions(
+                proxy.contract.functions.contract_version().call(),
+                EXPECTED_CONTRACTS_VERSION,
+            )
+            if not is_valid_version:
+                raise ContractVersionMismatch('Incompatible ABI for Discovery')
+        except BadFunctionCallOutput:
+            raise AddressWrongContract('')
```

Comment:
what about using the hash of the source code? since it is included as part of the bytecode, I assume it should be easy to get from the solc metadata and the deployed smart contract.

Reply:
>  using the hash of the source code?
> I assume it should be easy to get from the solc metadata and the deployed smart contract.
@hackaugusto 

Let me see if I understand you correctly. You are proposing to include the hash of the deployed contract sourcecode in the error here? What would that achieve?

Or maybe you are proposing to always check before each RPC call (or maybe only at the proxy ctor?) the deployed address sourcecode and see if it matches our expected deployed sourcecode?

Reply:
> You are proposing to include the hash of the deployed contract sourcecode in the error here? What would that achieve?

No. This PR is relying on the function `contract_version` to validate the bytecode, which may fail on some instances. Therefor I suggested an alternative.

With the above clarified, my suggestion didn't make much sense. I thought the hash of original the source code was appended at the end of the bytecode, but it's actually the [hash of the metadata](https://solidity.readthedocs.io/en/v0.4.24/metadata.html?highlight=hash#encoding-of-the-metadata-hash-in-the-bytecode).

Forget about the swarm hash, another alternative is to add the hash of the bytecode ourselves in the settings. E.g.:

```sh
$ solc raiden=contracts --combined-json bin contracts/TokenNetwork.sol 2>/dev/null | jq '.["contracts"]["contracts/TokenNetwork.sol:TokenNetwork"]["bin"]' | python -c 'import sys,raiden.utils,binascii; print(binascii.hexlify(raiden.utils.sha3(binascii.unhexlify(sys.stdin.read()[1:-2]))))'
b'1a1c260c1c1c9c61b3a5b916af3da7360b42f467fd3827db248ae1af7b8dad30'
```

And then using [`eth_getCode`](https://github.com/ethereum/wiki/wiki/JSON-RPC#eth_getcode)

Reply:
Yes I can do that. But this would need to be taken from the verified contract code (in etherscan) as apparently even a different way of combining the contracts with different pragmas as @ulope found out while trying to verify them can lead to different hash.

Reply:
I don't understand what etherscan has to do with this feature.

Reply:
I was writing a big reply but I managed to close my browser window and it did not save the reply.

So tl;dr:
We have had a lot of problems trying to verify contracts in etherscan and I fear similar problems here if we try to also have this features for the tests and the hash autogenerated by the user's system.

But if we do it ourselves only once for each deployment, save the bytecode hashes in constants.py and just skip the check if we are not connected to ropsten and later mainnet it should be fine!

But from some small experiments at my computer I still don't get the same bytecode.
@ulope Any ideas?

- **Local raiden node connected to ropsten**

` hexlify(sha3(jsonrpc_client.web3.eth.getCode(to_checksum_address(registry_address))))`                                                                                                                                                                                   
> b'276354fc2122512dc63e08457a5c00723e8c5a2d83c2a6758c13d4262543b459'

- **Solc v0.4.24 with the same solc arguments and source as the deployed contract (copy paste from etherscan verified source)**

`solc raiden=contracts --combined-json bin ~/TEST.sol 2>/dev/null | jq '.["contracts"]["/home/lefteris/TEST.sol:TokenNetwork"]["bin"]' | python -c 'import sys,raiden.utils,binascii; print(binascii.hexlify(raiden.utils.sha3(binascii.unhexlify(sys.stdin.read()[1:-2]))))'`
> b'ca7180531f2a0ec674fe9765e06bef09e48f2b65fd319541945a98663bdda50a'